Basic/DyanmicPrefabs/Assets/Scripts/01_Server authoritative dynamic prefab loading across peers/DynamicPrefabManager.cs

```diff
@@ -0,0 +1,476 @@
+        /// All spawned objects are invisible to clients that don't have the prefab loaded.
+        /// The server tells the clients that lack the preloaded prefab to load it and acknowledge that they've loaded it,
+        /// and then the server makes the object visible to that client.
+        /// </summary>
+        /// <param name="guid"></param>
+        /// <returns></returns>
+        public async Task<NetworkObject> SpawnWithVisibilitySystem(string guid, Vector3 position, Quaternion rotation)
+        {
+            if (IsServer)
+            {
+                var assetGuid = new AddressableGUID()
+                {
+                    Value = guid
+                };
+                
+                return await Spawn(assetGuid);
+            }
+
+            return null;
+
+            async Task<NetworkObject> Spawn(AddressableGUID assetGuid)
+            {
+                var prefab = await LoadDynamicPrefab(assetGuid);
+                var obj = Instantiate(prefab, position, rotation).GetComponent<NetworkObject>();
+                
+                if(m_PrefabHashToNetworkObjectId.TryGetValue(assetGuid.GetHashCode(), out var networkObjectIds))
+                {
+                    networkObjectIds.Add(obj);
+                }
+                else
+                {
+                    m_PrefabHashToNetworkObjectId.Add(assetGuid.GetHashCode(), new HashSet<NetworkObject>() {obj});
+                }
+
+                obj.CheckObjectVisibility = (clientId) => 
+                {
+                    //if the client has already loaded the prefab - we can make the object visible to them
+                    if (HasClientLoadedPrefab(clientId, assetGuid.GetHashCode()))
+                    {
+                        return true;
+                    }
+                    //otherwise the clients need to load the prefab, and after they ack - the ShowHiddenObjectsToClient 
+                    LoadAddressableClientRpc(assetGuid, new ClientRpcParams(){Send = new ClientRpcSendParams(){TargetClientIds = new ulong[]{clientId}}});
+                    return false;
+                };
+                
+                obj.SpawnWithOwnership(m_NetworkManager.LocalClientId);
+
+                return obj;
+            }
+        }
+
+        [ClientRpc]
+        void LoadAddressableClientRpc(AddressableGUID guid, ClientRpcParams rpcParams = default)
+        {
+            if (!IsHost)
+            {
+                Load(guid);
+            }
+
+            async void Load(AddressableGUID assetGuid)
+            {
+                Debug.Log("Loading dynamic prefab on the client...");
+                await LoadDynamicPrefab(assetGuid);
+                Debug.Log("Client loaded dynamic prefab");
+                AcknowledgeSuccessfulPrefabLoadServerRpc(assetGuid.GetHashCode());
+            }
+        }
+
+        [ServerRpc(RequireOwnership = false)]
```

Comment:
With require ownership false, could a malicious player ack an asset for another player and mess with that player's spawning and potentially bork that player's client?

Reply:
Oh boy, that's a good question. Let's theorise for a bit: is it possible to fake the originating client id for an rpc call?

Reply:
no clue, that'd be an NGO/UTP question. And might be out of scope for now thinking of it.